File: codebase/src/cpp/test13/federation/DestroyFederationTest.cpp
/*
 *   Copyright 2007 The Portico Project
 *
 *   This file is part of portico.
 *
 *   portico is free software; you can redistribute it and/or modify
 *   it under the terms of the Common Developer and Distribution License (CDDL)
 *   as published by Sun Microsystems. For more information see the LICENSE file.
 *
 *   Use of this software is strictly AT YOUR OWN RISK!!!
 *   If something bad happens you do not have permission to come crying to me.
 *   (that goes for your lawyer as well)
 *
 */
#include "DestroyFederationTest.h"

// Register test suite with the global repository
CPPUNIT_TEST_SUITE_REGISTRATION( DestroyFederationTest );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( DestroyFederationTest, "DestroyFederationTest" );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( DestroyFederationTest, "destroy" );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( DestroyFederationTest, "federationManagement" );

/////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////// Constructors/Destructors /////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////
DestroyFederationTest::DestroyFederationTest()
{
}

DestroyFederationTest::~DestroyFederationTest()
{
}

/////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////// Test Setup and Helper Methods ///////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////

void DestroyFederationTest::setUp()
{
    this->defaultFederate = new TestNG6Federate( "defaultFederate" );
    this->defaultFederate->quickCreate();
}

void DestroyFederationTest::tearDown()
{
	this->defaultFederate->quickDestroyNoFail();
	delete this->defaultFederate;
}

////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////// Test Methods ///////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////

///////////////////////////////////////////
// TEST: (valid) testDestroyFederation() //
///////////////////////////////////////////
void DestroyFederationTest::testDestroyFederation()
{
	// try and destroy the federation
	try
	{
		defaultFederate->rtiamb->destroyFederationExecution( TestNG6Federate::SIMPLE_NAME );
	}
	catch( RTI::Exception &e )
	{
		// failure
		failTest( "Unexpected exception while destroying federation: (%s) %s", e._name, e._reason );
	}
}

////////////////////////////////////////////////////
// TEST: testDestroyFederationWithJoinedMembers() //
////////////////////////////////////////////////////
void DestroyFederationTest::testDestroyFederationWithJoinedMembers()
{
	// join the federation so that there are still federates joined when we
	// try to destroy it
	defaultFederate->quickJoin();
	
	// try and destroy the federation
	try
	{
		defaultFederate->rtiamb->destroyFederationExecution( TestNG6Federate::SIMPLE_NAME );
		defaultFederate->quickResign();
		failTestMissingException( "FederatesCurrentlyJoined",
		                          "destroying fderation with joined members" );
	}
	catch( RTI::FederatesCurrentlyJoined &fcj )
	{
		// success!
		defaultFederate->quickResign();
	}
	catch( RTI::Exception &e )
	{
		// failure
		defaultFederate->quickResign();
		failTestWrongException( "FederatesCurrentlyJoined", e,
		                        "destroying federation with joined members" );
	}
}

///////////////////////////////////////////////////
// TEST: testDestroyFederationThatDoesNotExist() //
///////////////////////////////////////////////////
void DestroyFederationTest::testDestroyFederationThatDoesNotExist()
{
	// try and destroy the federation
	try
	{
		defaultFederate->rtiamb->destroyFederationExecution( "NoSuchFederation" );
		failTestMissingException( "FederationExecutionDoesNotExist",
		                          "destroying federation that doesn't exist" );
	}
	catch( RTI::FederationExecutionDoesNotExist &fedne )
	{
		// success!
	}
	catch( RTI::Exception &e )
	{
		// failure
		failTestWrongException( "FederationExecutionDoesNotExist", e,
		                        "destroying federation that doesn't exist" );
	}
}

///////////////////////////////////////////////
// TEST: testDestroyFederationWithNullName() //
///////////////////////////////////////////////
void DestroyFederationTest::testDestroyFederationWithNullName()
{
	// try and destroy the federation
	try
	{
		defaultFederate->rtiamb->destroyFederationExecution( NULL );
		failTestMissingException( "FederationExecutionDoesNotExist",
		                          "destroying federation with null name" );
	}
	catch( RTI::FederationExecutionDoesNotExist &fedne )
	{
		// success!
	}
	catch( RTI::Exception &e )
	{
		// failure
		failTestWrongException( "FederationExecutionDoesNotExist", e,
		                        "destroying federation with null name" );
	}
}

File: codebase/src/java/test/hlaunit/ieee1516/federation/DestroyFederationTest.java
/*
 *   Copyright 2007 The Portico Project
 *
 *   This file is part of portico.
 *
 *   portico is free software; you can redistribute it and/or modify
 *   it under the terms of the Common Developer and Distribution License (CDDL) 
 *   as published by Sun Microsystems. For more information see the LICENSE file.
 *   
 *   Use of this software is strictly AT YOUR OWN RISK!!!
 *   If something bad happens you do not have permission to come crying to me.
 *   (that goes for your lawyer as well)
 *
 */
package hlaunit.ieee1516.federation;

import hla.rti1516.FederatesCurrentlyJoined;
import hla.rti1516.FederationExecutionDoesNotExist;
import hla.rti1516.RTIinternalError;
import hlaunit.ieee1516.common.Abstract1516Test;

import org.testng.Assert;
import org.testng.annotations.AfterClass;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

@Test(singleThreaded=true, groups={"DestroyFederationTest", "basic", "destroy", "federationManagement"})
public class DestroyFederationTest extends Abstract1516Test
{
	//----------------------------------------------------------
	//                    STATIC VARIABLES
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                   INSTANCE VARIABLES
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                      CONSTRUCTORS
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                    INSTANCE METHODS
	//----------------------------------------------------------
	@BeforeClass(alwaysRun=true)
	public void beforeClass()
	{
		super.beforeClass();
	}

	@BeforeMethod(alwaysRun=true)
	public void beforeMethod()
	{
		defaultFederate.quickCreate();
	}
	
	@AfterMethod(alwaysRun=true)
	public void afterMethod()
	{
		defaultFederate.quickDestroyTolerant();
	}
	
	@AfterClass(alwaysRun=true)
	public void afterClass()
	{
		super.afterClass();
	}

	///////////////////////////////////////////
	// TEST: (valid) testDestroyFederation() //
	///////////////////////////////////////////
	@Test
	public void testDestroyFederation()
	{
		try
		{
			// destroy the federation //
			defaultFederate.rtiamb.destroyFederationExecution( defaultFederate.simpleName );
		}
		catch( Exception e )
		{
			Assert.fail( "Failed to destroy existing federation", e );
		}
		
		try
		{
			// ensure that federation is gone and can't be destroyed again //
			defaultFederate.rtiamb.destroyFederationExecution( defaultFederate.simpleName );
			Assert.fail( "Could not ensure that valid federation was destroyed" );
		}
		catch( FederationExecutionDoesNotExist dne )
		{
			// SUCCESS!
		}
		catch( Exception e )
		{
			Assert.fail( "Failed to destroy existing federation", e );
		}
	}
	
	///////////////////////////////////////////////////
	// TEST: testDestroyFederationThatDoesNotExist() //
	///////////////////////////////////////////////////
	@Test
	public void testDestroyFederationThatDoesNotExist()
	{
		// try and destroy a federation that does not exist //
		try
		{
			defaultFederate.rtiamb.destroyFederationExecution( "noSuchFederation" );
			Assert.fail( "No exception while destroying non-existent federation" );
		}
		catch( FederationExecutionDoesNotExist dne )
		{
			// SUCCESS
		}
		catch( Exception e )
		{
			Assert.fail( "Wrong exception while destorying non-existent federation", e );
		}
	}
	
	///////////////////////////////////////////////
	// TEST: testDestroyFederationWithNullName() //
	///////////////////////////////////////////////
	@Test
	public void testDestroyFederationWithNullName()
	{
		// try and destroy with null federation name //
		try
		{
			defaultFederate.rtiamb.destroyFederationExecution( null );
			Assert.fail( "No exception while destroying federation with null name" );
		}
		catch( FederationExecutionDoesNotExist ne )
		{
			// SUCCESS
		}
		catch( RTIinternalError rtie )
		{
			// ALSO FINE
		}
		catch( Exception e )
		{
			Assert.fail( "Wrong exception while destorying federation with null name", e );
		}
	}
	
	//////////////////////////////////////////////////////
	// TEST: testDestroyFederationWithJoinedFederates() //
	//////////////////////////////////////////////////////
	@Test
	public void testDestroyFederationWithJoinedFederates()
	{
		// join the federation
		defaultFederate.quickJoin();
		
		try
		{
			// destroy the federation //
			defaultFederate.rtiamb.destroyFederationExecution( defaultFederate.simpleName );
			Assert.fail( "Was able to destroy a federation with active federates" );
		}
		catch( FederatesCurrentlyJoined fcj )
		{
			// SUCCESS
		}
		catch( Exception e )
		{
			Assert.fail( "Wrong exception while destroying federation with active federates", e );
		}
		finally
		{
			defaultFederate.quickResign();
			defaultFederate.quickDestroy();
		}
	}
	
	//----------------------------------------------------------
	//                     STATIC METHODS
	//----------------------------------------------------------

}
